doc/design/logging.md

```diff
@@ -0,0 +1,396 @@
+
+    return 0;
+}
+```
+
+#### Add logging support for custom types
+
+```cpp
+#include "iceoryx_hoofs/log/logging.hpp"
+#include "iceoryx_hoofs/log/logstream.hpp"
+#include <cstdint>
+
+struct MyType
+{
+    uint16_t a {0};
+    uint16_t b {0};
+};
+
+iox::log::LogStream& operator<<(iox::log::LogStream& stream, const MyType& m)
+{
+    stream << "MyType { a = " << m.a << "; b = " << m.b << " }";
+    return stream;
+}
+
+int main()
+{
+    MyType m;
+    IOX_LOG(INFO) << m;
+
+    return 0;
+}
+```
+
+#### Creating a custom logger
+
+This is a simple custom logger which prints the log level as emoji and ignores
+the file, line number and function name.
+
+```cpp
+#include "iceoryx_hoofs/log/logger.hpp"
+#include "iceoryx_hoofs/log/logging.hpp"
+
+class MyLogger : public iox::log::Logger
+{
+  public:
+    static void init()
+    {
+        static MyLogger myLogger;
+        iox::log::setActiveLogger(&myLogger);
+        iox::log::Logger::init(iox::log::logLevelFromEnvOr(iox::log::LogLevel::INFO));
+    }
+
+  private:
+    void setupNewLogMessage(const char*, const int, const char*, iox::log::LogLevel logLevel) override
+    {
+        switch(logLevel) {
+            case iox::log::LogLevel::FATAL:
+                logString("💀: ");
+                break;
+            case iox::log::LogLevel::ERROR:
+                logString("🙈: ");
+                break;
+            case iox::log::LogLevel::WARN:
+                logString("🙀: ");
+                break;
+            case iox::log::LogLevel::INFO:
+                logString("💘: ");
+                break;
+            case iox::log::LogLevel::DEBUG:
+                logString("🐞: ");
+                break;
+            case iox::log::LogLevel::TRACE:
+                logString("🐾: ");
+                break;
+            default:
+                logString("🐔: ");
+        }
+    }
+
+    void flush() override {
+        puts(iox::log::Logger::buffer());
+        iox::log::Logger::assumePushed();
+    }
+};
+
+int main()
+{
+    MyLogger::init();
+
+    IOX_LOG(FATAL) << "Whoops ... look, over there is a dead seagull flying!";
+    IOX_LOG(ERROR) << "Oh no!";
+    IOX_LOG(WARN) << "It didn't happen!";
+    IOX_LOG(INFO) << "All glory to the hypnotoad!";
+    IOX_LOG(DEBUG) << "I didn't do it!";
+    IOX_LOG(TRACE) << "Row row row your boat!";
+
+    return 0;
+}
+```
+
+## Open issues
+
+- do we need to change the log level after `Logger::init`
+- do we want a `IOX_LOG_IF(cond, level)` macro
+- shall the TestingLogger register signals to catch SIGTERM, etc. and print the
+  log messages when the signal is raised? It might be necessary to wait for the
+  error handling refactoring before this can be done
+- instead of having the `Logger::init()` static function with hidden default
+  parameter this could be replaced by `Logger::init(LogLevel::WARN)`,
+  `Logger::initFromEnvOr(LogLevel::WARN)` and a builder like
+  `Logger::customize().logLevelFromEnvOr(LogLevel::WARN).init()`
```

Comment:
I think I will try this when the class is implemented and if it feels nice I will also update the design document. What do you think?

Reply:
I assume this refers to the \_\_FILE\_\_ \_\_LINE\_\_ etc.? Sure, not critical at all and easy to refactor.